influxdb3_write/src/wal.rs

```diff
@@ -1,44 +1,642 @@
 //! This is the implementation of the `Wal` that the buffer uses to make buffered data durable
 //! on disk.
 
-use crate::{SegmentFile, SegmentId, Wal, WalSegmentReader, WalSegmentWriter};
+use crate::{
+    SegmentFile, SegmentId, SegmentIdBytes, SequenceNumber, Wal, WalOp, WalOpBatch,
+    WalSegmentReader, WalSegmentWriter,
+};
 use async_trait::async_trait;
-use std::path::PathBuf;
-use std::sync::Arc;
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use crc32fast::Hasher;
+use datafusion::parquet::file::reader::Length;
+use observability_deps::tracing::{info, warn};
+use snap::read::FrameDecoder;
+use std::fmt::Debug;
+use std::{
+    fs::{File, OpenOptions},
+    io::{self, BufReader, Cursor, Read, Write},
+    mem,
+    path::PathBuf,
+};
+use thiserror::Error;
+
+/// The first bytes written into a segment file to identify it and its version.
+type FileTypeIdentifier = [u8; 8];
+const FILE_TYPE_IDENTIFIER: &[u8] = b"idb3.001";
+
+/// File extension for segment files
+const SEGMENT_FILE_EXTENSION: &str = "wal";
```

Comment:
I think I might have been confused about our previous offline conversations. Are the segment files and wal one and the same? My understanding was that the wal just contained data that had not been persisted yet (with maybe 1 or more segments of nonpersisted data) and a segment that was persisted to disk was a blob of data containing the list of all the parquet files for that segment. To me the terminology feels a bit fuzzy and intermingled.

Reply:
The wal contains all data that is in the in-memory buffer for a given segment. So there are a few different things here:
* A Buffer Segment (in memory collection of writes)
* A WAL Segment (a file on locally attached disk that has the durable record of what is in a buffer segment)
* A Segment File (a file in object store that has the summary information of what parquet files were persisted for a given buffer segment)

The Segment File could arguably be renamed to something more like `segment_persist_info` or something like that.

Reply:
Okay that makes more sense to me. I think when I work on the persister I'll give Segment Files a name like that to be more clear